### PR TITLE
final_video.py: ffmpeg-python port

### PR DESCRIPTION
**Pull request by Xpl0itU**
# Description
Uses ffmpeg-python, which improves video rendering performance by a lot, it can render a video in less than a minute in my system

# Checklist:

- [X] I am pushing changes to the **develop** branch
- [ ] I am using the recommended development environment
- [X] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have formatted and linted my code using python-black and pylint
- [X] I have cleaned up unnecessary files
- [X] My changes generate no new warnings
- [ ] My changes follow the existing code-style
- [X] My changes are relevant to the project

# Any other information (e.g how to test the changes)

None

### CODE REVIEW

**Comment by JasonLovesDoggo** [comment]
any updates?

**Comment by OpenSourceSimon** [comment]
> Any updates?

There is someone else working on this too, the original creator can't solve the to-dos:
- Story mode
- Watermark
- Custom resolution (especially landscape)

**Comment by OpenSourceSimon** [comment]
Update on the progress: almost everything is done! Story mode and landscape is fixed, only thing that's left to do is the watermark implementation. We'll keep you notified if there are any new updates

**Comment by JasonLovesDoggo** [comment]
Lmk when completed

File: video_creation/final_video.py [hunk 1]
```diff
@@ -1,35 +1,67 @@
...
 console = Console()
 
+import tempfile
+import threading
+import time
+
+
+class ProgressFfmpeg(threading.Thread):
+    def __init__(self, vid_duration_seconds, progress_update_callback):
+        threading.Thread.__init__(self, name='ProgressFfmpeg')
```
Comment (electro199) [inline video_creation/final_video.py:30]: always use super.__init__()

File: video_creation/final_video.py [hunk 2]
```diff
@@ -1,35 +1,67 @@
...
+        self.progress_update_callback = progress_update_callback
+
+    def run(self):
+
+        while not self.stop_event.is_set():
+            latest_progress = self.get_latest_ms_progress()
+            if latest_progress is not None:
+                completed_percent = latest_progress / self.vid_duration_seconds
+                self.progress_update_callback(completed_percent)
+            time.sleep(1)
```
Comment (electro199) [inline video_creation/final_video.py:43]: why sleep here

File: video_creation/final_video.py [hunk 3]
```diff
@@ -209,18 +224,42 @@ def make_final_video(
...
+    def on_update_example(progress):
+        status = round(progress * 100, 2)
+        old_percentage = pbar.n
+        pbar.update(status - old_percentage)
+
+    position = W // 2, H - 20
+    ffmpeg.filter(filter_name='drawtext', stream_spec=background_clip, text=text, fontfile='fonts/Roboto-Regular.ttf', fontsize=12,
+                              fontcolor='white', x=position[0], y=position[1], box=1)
+    with ProgressFfmpeg(length, on_update_example) as progress:
+        ffmpeg.output(background_clip, audio, f"results/{subreddit}/{filename}.mp4", f='mp4',
```
Comment (electro199) [inline video_creation/final_video.py:252]: `"results/{subreddit}/{filename[:250]}.mp4`

File: video_creation/final_video.py [hunk 4]
```diff
@@ -1,33 +1,67 @@
...
-from utils.video import Video
-from utils.videos import save_data
-from utils.thumbnail import create_thumbnail
-from utils import settings
 from utils.thumbnail import create_thumbnail
+from utils.videos import save_data
 
 console = Console()
 
+import tempfile
```
Comment (electro199) [inline video_creation/final_video.py:23]: from where tempfile is comming from ?

Comment (OpenSourceSimon) [reply video_creation/final_video.py:23]: Built-in python module

File: utils/.config.template.toml [hunk 5]
```diff
@@ -29,7 +29,6 @@ transition = { optional = true, default = 0.2, example = 0.2, explanation = "Set
 storymode = { optional = true, type = "bool", default = false, example = false, options = [true, false,], explanation = "Only read out title and post content, great for subreddits with stories" }
 storymodemethod= { optional = true, default = 1, example = 1, explanation = "Style that's used for the storymode. Set to 0 for single picture display in whole video, set to 1 for fancy looking video ", type = "int", nmin = 0, oob_error = "It's very hard to run something less than once.", options = [0, 1] }
 storymode_max_length = { optional = true, default = 1000, example = 1000, explanation = "Max length of the storymode video in characters. 200 characters are approximately 50 seconds.", type = "int", nmin = 1, oob_error = "It's very hard to make a video under a second." }
-fps = { optional = false, default = 30, example = 30, explanation = "Sets the FPS of the video, 30 is default for best performance. 60 FPS is smoother.", type = "int", nmin = 1, nmax = 60, oob_error = "The FPS HAS to be between 1 and 60" }
```
Comment (electro199) [inline utils/.config.template.toml:32]: why removed ?

Comment (OpenSourceSimon) [reply utils/.config.template.toml:32]: Because we haven't implemented it with ffmpeg and honestly nobody uses it

Comment (electro199) [reply utils/.config.template.toml:32]: ok make sense

File: video_creation/final_video.py (hunk 3 above)
Comment (OpenSourceSimon) [reply video_creation/final_video.py:257]:
```suggestion
        ffmpeg.output(background_clip, audio, f"results/{subreddit}/{filename[:250]}.mp4", f='mp4',
```

**Review by OpenSourceSimon — APPROVED** [review]

**Pull request merged**